### ISSUE TITLE
The Cilium agent daemon creation failed, because  installing iptables rules failed.

### ISSUE THREAD

**Issue by dwbxm** [cilium/cilium#27204]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

Cilium agent daemon creation failed when I install cilium on kubernetes without kube-porxy. The details are as follows：
1. cmd：
helm install cilium cilium/cilium --version 1.13.4 \
    --namespace kube-system \
    --set kubeProxyReplacement=strict \
    --set k8sServiceHost=${API_SERVER_IP} \
    --set k8sServicePort=${API_SERVER_PORT}

2. Cilium status
cilium-vfm4f  unable to retrieve cilium status: container cilium-agent is in CrashLoopBackOff, exited with code 1: level=fatal msg="failed to start: daemon creation failed: error while initializing daemon: failed while reinitializing datapath: failed to install iptables rules: unable to run 'iptables -t mangle -D OLD_CILIUM_PRE_mangle -m sock

[cilium-status.txt](https://github.com/cilium/cilium/files/12236992/cilium-status.txt)

What is the reason？

### Cilium Version

cilium-cli: v0.12.4 compiled with go1.19.1 on linux/amd64
cilium image (default): v1.12.2

### Kernel Version

Linux master1 6.2.7-1.el7.elrepo.x86_64 #1 SMP PREEMPT_DYNAMIC Thu Mar 16 14:05:13 EDT 2023 x86_64 x86_64 x86_64 GNU/Lin

### Kubernetes Version

Client Version: version.Info{Major:"1", Minor:"25", GitVersion:"v1.25.2", GitCommit:"5835544ca568b757a8ecae5c153f317e5736700e", GitTreeState:"clean", BuildDate:"2022-09-21T14:33:49Z", GoVersion:"go1.19.1", Compiler:"gc", Platform:"linux/amd64"}
Kustomize Version: v4.5.7
Server Version: version.Info{Major:"1", Minor:"25", GitVersion:"v1.25.2", GitCommit:"5835544ca568b757a8ecae5c153f317e5736700e", GitTreeState:"clean", BuildDate:"2022-09-21T14:27:13Z", GoVersion:"go1.19.1", Compiler:"gc", Platform:"linux/amd64"}

### Sysdump

[cilium-sysdump-20230802-022011.zip](https://github.com/cilium/cilium/files/12237019/cilium-sysdump-20230802-022011.zip)

Collecting sysdump with cilium-cli version: v0.12.4, args: [sysdump]
🔍 Collecting Kubernetes nodes
🔍 Collect Kubernetes nodes
🔍 Collect Kubernetes version
🔍 Collecting Kubernetes pods
🔍 Collecting Kubernetes namespaces
🔍 Collecting Kubernetes events
🔍 Collecting Kubernetes pods summary
🔍 Collecting Kubernetes services
🔍 Collecting Kubernetes network policies
🔍 Collecting Cilium cluster-wide network policies
🔍 Collecting Cilium network policies
🔍 Collecting Cilium local redirect policies
🔍 Collecting Cilium egress NAT policies
🔍 Collecting Cilium identities
🔍 Collecting Cilium endpoints
🔍 Collecting Cilium nodes
🔍 Collecting Kubernetes endpoints
🔍 Collecting Ingresses
🔍 Collecting CiliumClusterwideEnvoyConfigs
🔍 Collecting Cilium etcd secret
🔍 Collecting the Cilium configuration
🔍 Collecting the Cilium daemonset(s)
🔍 Collecting the Hubble daemonset
🔍 Collecting the Hubble Relay configuration
🔍 Collecting the Hubble Relay deployment
🔍 Collecting the Hubble UI deployment
🔍 Collecting the 'clustermesh-apiserver' deployment
⚠️ Deployment "hubble-relay" not found in namespace "kube-system" - this is expected if Hubble is not enabled
🔍 Collecting the Cilium operator deployment
🔍 Collecting the CNI configmap
🔍 Collecting the CNI configuration files from Cilium pods
🔍 Collecting gops stats from Cilium pods
🔍 Collecting gops stats from Hubble Relay pods
🔍 Collecting bugtool output from Cilium pods
🔍 Collecting logs from Cilium operator pods
🔍 Collecting logs from Cilium pods
🔍 Collecting logs from 'clustermesh-apiserver' pods
🔍 Collecting logs from Hubble pods
⚠️ Deployment "hubble-ui" not found in namespace "kube-system" - this is expected if Hubble UI is not enabled
⚠️ Deployment "clustermesh-apiserver" not found in namespace "kube-system" - this is expected if 'clustermesh-apiserver' isn't enabled
🔍 Collecting logs from Hubble UI pods
🔍 Collecting gops stats from Hubble pods
🔍 Collecting CiliumEnvoyConfigs
🔍 Collecting logs from Hubble Relay pods
🔍 Collecting bugtool output from Tetragon pods
🔍 Collecting platform-specific data
level=warning msg="Ignoring invalid CiliumEnvoyConfig JSON (proto: (line 1:405): unable to resolve \"type.googleapis.com/envoy.extensions.filters.http.router.v3.Router\": \"not found\"): {\n\t\"@type\": \"type.googleapis.com/envoy.config.listener.v3.Listener\",\n\t\"filterChains\": [\n\t\t{\n\t\t\t\"filterChainMatch\": {\n\t\t\t\t\"transportProtocol\": \"raw_buffer\"\n\t\t\t},\n\t\t\t\"filters\": [\n\t\t\t\t{\n\t\t\t\t\t\"name\": \"envoy.filters.network.http_connection_manager\",\n\t\t\t\t\t\"typedConfig\": {\n\t\t\t\t\t\t\"@type\": \"type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager\",\n\t\t\t\t\t\t\"httpFilters\": [\n\t\t\t\t\t\t\t{\n\t\t\t\t\t\t\t\t\"name\": \"envoy.filters.http.router\",\n\t\t\t\t\t\t\t\t\"typedConfig\": {\n\t\t\t\t\t\t\t\t\t\"@type\": \"type.googleapis.com/envoy.extensions.filters.http.router.v3.Router\"\n\t\t\t\t\t\t\t\t}\n\t\t\t\t\t\t\t}\n\t\t\t\t\t\t],\n\t\t\t\t\t\t\"rds\": {\n\t\t\t\t\t\t\t\"routeConfigName\": \"listener-insecure\"\n\t\t\t\t\t\t},\n\t\t\t\t\t\t\"statPrefix\": \"listener-insecure\",\n\t\t\t\t\t\t\"upgradeConfigs\": [\n\t\t\t\t\t\t\t{\n\t\t\t\t\t\t\t\t\"upgradeType\": \"websocket\"\n\t\t\t\t\t\t\t}\n\t\t\t\t\t\t],\n\t\t\t\t\t\t\"useRemoteAddress\": true\n\t\t\t\t\t}\n\t\t\t\t}\n\t\t\t]\n\t\t}\n\t],\n\t\"listenerFilters\": [\n\t\t{\n\t\t\t\"name\": \"envoy.filters.listener.tls_inspector\",\n\t\t\t\"typedConfig\": {\n\t\t\t\t\"@type\": \"type.googleapis.com/envoy.extensions.filters.listener.tls_inspector.v3.TlsInspector\"\n\t\t\t}\n\t\t}\n\t],\n\t\"name\": \"listener\",\n\t\"socketOptions\": [\n\t\t{\n\t\t\t\"description\": \"Enable TCP keep-alive (default to enabled)\",\n\t\t\t\"intValue\": \"1\",\n\t\t\t\"level\": \"1\",\n\t\t\t\"name\": \"9\",\n\t\t\t\"state\": \"STATE_LISTENING\"\n\t\t},\n\t\t{\n\t\t\t\"description\": \"TCP keep-alive idle time (in seconds) (defaults to 10s)\",\n\t\t\t\"intValue\": \"10\",\n\t\t\t\"level\": \"6\",\n\t\t\t\"name\": \"4\",\n\t\t\t\"state\": \"STATE_LISTENING\"\n\t\t},\n\t\t{\n\t\t\t\"description\": \"TCP keep-alive probe intervals (in seconds) (defaults to 5s)\",\n\t\t\t\"intValue\": \"5\",\n\t\t\t\"level\": \"6\",\n\t\t\t\"name\": \"5\",\n\t\t\t\"state\": \"STATE_LISTENING\"\n\t\t},\n\t\t{\n\t\t\t\"description\": \"TCP keep-alive probe max failures.\",\n\t\t\t\"intValue\": \"10\",\n\t\t\t\"level\": \"6\",\n\t\t\t\"name\": \"6\",\n\t\t\t\"state\": \"STATE_LISTENING\"\n\t\t}\n\t]\n}" subsys=k8s-apis-cilium.io-v2
🔍 Collecting kvstore data
🔍 Collecting Hubble flows from Cilium pods
⚠️ Container "cilium-agent" for pod "cilium-8bckz" in namespace "kube-system" is not running. Trying EphemeralContainer or separate Pod instead...
⚠️ Container "cilium-agent" for pod "cilium-bbd4n" in namespace "kube-system" is not running. Trying EphemeralContainer or separate Pod instead...
⚠️ Container "cilium-agent" for pod "cilium-bmnb4" in namespace "kube-system" is not running. Trying EphemeralContainer or separate Pod instead...
⚠️ Container "cilium-agent" for pod "cilium-vfm4f" in namespace "kube-system" is not running. Trying EphemeralContainer or separate Pod instead...
⚠️ EphemeralContainer "sysdump-1690957213" on pod "cilium-8bckz" in namespace "kube-system" never reached Running status (falling back to separate Pod)

### Relevant log output

```shell
The logs in /var/log/messages:
Aug  2 02:55:00 master1 kubelet: E0802 02:55:00.820417   26050 pod_workers.go:965] "Error syncing pod, skipping" err="failed to \"StartContainer\" for \"cilium-agent\" with CrashLoopBackOff: \"back-off 5m0s restarting failed container=cilium-agent pod=cilium-8bckz_kube-system(1ad22c6d-40a2-4517-9e83-8b333d9f0d15)\"" pod="kube-system/cilium-8bckz" podUID=1ad22c6d-40a2-4517-9e83-8b333d9f0d15
Aug  2 02:55:13 master1 kubelet: E0802 02:55:13.820882   26050 pod_workers.go:965] "Error syncing pod, skipping" err="failed to \"StartContainer\" for \"cilium-agent\" with CrashLoopBackOff: \"back-off 5m0s restarting failed container=cilium-agent pod=cilium-8bckz_kube-system(1ad22c6d-40a2-4517-9e83-8b333d9f0d15)\"" pod="kube-system/cilium-8bckz" podUID=1ad22c6d-40a2-4517-9e83-8b333d9f0d15
Aug  2 02:55:28 master1 kubelet: E0802 02:55:28.821211   26050 pod_workers.go:965] "Error syncing pod, skipping" err="failed to \"StartContainer\" for \"cilium-agent\" with CrashLoopBackOff: \"back-off 5m0s restarting failed container=cilium-agent pod=cilium-8bckz_kube-system(1ad22c6d-40a2-4517-9e83-8b333d9f0d15)\"" pod="kube-system/cilium-8bckz" podUID=1ad22c6d-40a2-4517-9e83-8b333d9f0d15
Aug  2 02:55:43 master1 kubelet: E0802 02:55:43.819358   26050 pod_workers.go:965] "Error syncing pod, skipping" err="failed to \"StartContainer\" for \"cilium-agent\" with CrashLoopBackOff: \"back-off 5m0s restarting failed container=cilium-agent pod=cilium-8bckz_kube-system(1ad22c6d-40a2-4517-9e83-8b333d9f0d15)\"" pod="kube-system/cilium-8bckz" podUID=1ad22c6d-40a2-4517-9e83-8b333d9f0d15
Aug  2 02:55:55 master1 containerd: time="2023-08-02T02:55:55.251829088-04:00" level=info msg="ImageCreate event &ImageCreate{Name:quay.io/cilium/operator-generic@sha256:88429d3e5d05e3805919a2958e002158e26da4ba73e4a8c894e2d5af066136c4,Labels:map[string]string{io.cri-containerd.image: managed,},XXX_unrecognized:[],}"
Aug  2 02:55:55 master1 containerd: time="2023-08-02T02:55:55.252127186-04:00" level=error msg="Failed to handle backOff event &ImageCreate{Name:quay.io/cilium/operator-generic@sha256:88429d3e5d05e3805919a2958e002158e26da4ba73e4a8c894e2d5af066136c4,Labels:map[string]string{io.cri-containerd.image: managed,},XXX_unrecognized:[],} for quay.io/cilium/operator-generic@sha256:88429d3e5d05e3805919a2958e002158e26da4ba73e4a8c894e2d5af066136c4" error="update image store for \"quay.io/cilium/operator-generic@sha256:88429d3e5d05e3805919a2958e002158e26da4ba73e4a8c894e2d5af066136c4\": get image info from containerd: get image diffIDs: unexpected media type application/octet-stream for sha256:88429d3e5d05e3805919a2958e002158e26da4ba73e4a8c894e2d5af066136c4: not found"
Aug  2 02:55:58 master1 kubelet: E0802 02:55:58.818480   26050 pod_workers.go:965] "Error syncing pod, skipping" err="failed to \"StartContainer\" for \"cilium-agent\" with CrashLoopBackOff: \"back-off 5m0s restarting failed container=cilium-agent pod=cilium-8bckz_kube-system(1ad22c6d-40a2-4517-9e83-8b333d9f0d15)\"" pod="kube-system/cilium-8bckz" podUID=1ad22c6d-40a2-4517-9e83-8b333d9f0d15
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by dylandreimerink** [comment]
Thank you for the report! 

Did a bit of digging:
```
2023-08-02T02:17:12.055050257-04:00 level=fatal msg="failed to start: daemon creation failed: error while initializing daemon: failed while reinitializing datapath: failed to install iptables rules: unable to run 'iptables -t mangle -D OLD_CILIUM_PRE_mangle -m socket [unsupported revision] -m comment --comment cilium: any->pod redirect proxied traffic to host proxy -j MARK --set-xmark 0x200/0xffffffff' iptables command: exit status 2 stderr=\"Bad argument `[unsupported'\\nTry `iptables -h' or 'iptables --help' for more information.\\n\"" subsys=daemon
```

The part of the code that fails is attempting to delete old iptables rules. Obviously the `[unsupported revision]` part is causing the issues. The deleting works by first listing the existing rules then deleting them. The `[unsupported revision]` string is part of the rules echoed back by `iptables-save`, this is a snippet:

```
[0:0] -A CILIUM_PRE_mangle -p udp -m mark --mark 0x4b470200 -m comment --comment "cilium: TPROXY to host default/cilium-gateway-my-gateway/listener proxy" -j TPROXY --on-port 18251 --on-ip 0.0.0.0 --tproxy-mark 0x200/0xffffffff
[0:0] -A OLD_CILIUM_PRE_mangle -m socket [unsupported revision] -m comment --comment "cilium: any->pod redirect proxied traffic to host proxy" -j MARK --set-xmark 0x200/0xffffffff
[0:0] -A OLD_CILIUM_PRE_mangle -p tcp -m mark --mark 0xab910200 -m comment --comment "cilium: TPROXY to host cilium-dns-egress proxy" -j TPROXY --on-port 37291 --on-ip 0.0.0.0 --tproxy-mark 0x200/0xffffffff
```

Related IPTables commits for reference:
* https://git.netfilter.org/iptables/commit/?id=bfa3321fedcb5b4b313d255753351397447a5669
* https://git.netfilter.org/iptables/commit/?id=12a52ff9cc9944345204d3e429dd4414448fbcd2

If I understand it correctly, this happens when the rules in the kernel were created with a revision of the `socket` module that the iptables utility doesn't know. Therefor it can't print the rule as human readable because it doesn't know how to interpret the data. This apparently can happen when upgrading or downgrading iptables version.

**Comment by dwbxm** [comment]
Thanks!   @dylandreimerink 

Sorry, I missed this log message.
I  think your analysis is correct, and I will conduct another analysis and test based on this information. 

Thank you so much！

**Comment by dwbxm** [comment]
@dylandreimerink    Thanks.  I installed  the Cilium  version 1.13.2 .  Then I uninstalled  the old version with  cmd "cilium uninstalled"  ，and  helm uninstalled cilium -n kube-system(because cilium is not purely uninstalled after executing cmd "cilium uninstalled".
Then I installed the cilium version 1.13.4.  

Now, I suspect it is caused by the incorrect version installation, but I have not tested it.
